### PR TITLE
Change register to apply

### PR DESCRIPTION
**Pull request by zhilingc**
Change register API to apply. 

If you apply a spec for a resource that doesn't already exist in the system, it will be created. If not, then the resource will be updated to match the new spec provided. Note that only fields without dependencies can be updated; i.e. changes to ids, storage sinks, or value types are not allowed. If you apply a spec that has no changes without throwing an error.

Updated the CLI to reflect these changes as well.

### CODE REVIEW

File: core/src/main/java/feast/core/model/FeatureGroupInfo.java [hunk 1]
```diff
@@ -111,4 +111,23 @@ public FeatureGroupDetail getFeatureGroupDetail() {
...
             .build();
   }
+
+  /**
+   * Checks if this is eq to the other given feature group
+   *
+   * @param otherFeatureGroup
+   * @return boolean
+   */
+  public boolean eq(FeatureGroupInfo otherFeatureGroup) {
```
Comment (woop) [inline core/src/main/java/feast/core/model/FeatureGroupInfo.java]: Isn't it possible (and perhaps cleaner) to use the built in .equals() method that all Java objects have? 

https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#equals(java.lang.Object)

File: core/src/test/java/feast/core/model/EntityInfoTest.java [hunk 2]
```diff
@@ -66,4 +67,13 @@ public void shouldBuildAndReturnCorrespondingDetail() {
             EntityDetail.newBuilder().setSpec(entitySpec).setLastUpdated(ts).build();
     assertThat(entityInfo.getEntityDetail(), equalTo(expected));
   }
+
+  @Test
+  public void shouldBeEqualToEntityFromSameSpecs() {
+    EntityInfo entity1 = new EntityInfo(entitySpec);
+    entity1.setCreated(Date.from(Instant.ofEpochSecond(1)));
+    EntityInfo entity2 = new EntityInfo(entitySpec);
+    entity1.setCreated(Date.from(Instant.ofEpochSecond(2)));
```
Comment (woop) [inline core/src/test/java/feast/core/model/EntityInfoTest.java]: Is this line correct? Shouldn't it be `entity2.setCreated(Date.from(Instant.ofEpochSecond(2)));`?

File: core/src/main/java/feast/core/model/FeatureGroupInfo.java [hunk 3]
```diff
@@ -63,68 +62,75 @@ public FeatureGroupInfo() {
...
+    return isStoreEqual(this.warehouseStore, this.warehouseStoreOpts, updatedWarehouseStore)
+        && isStoreEqual(this.servingStore, this.servingStoreOpts, updatedServingStore);
+  }
+
+  private boolean isStoreEqual(StorageInfo oldStore, String oldStoreOpts, DataStore newStore) {
+    return getStorageId(oldStore).equals(newStore == null ? "" : newStore.getId())
+        && oldStoreOpts.equals(
+            newStore == null
+                ? ""
+                : TypeConversion.convertMapToJsonString(newStore.getOptionsMap()));
```
Comment (woop) [inline core/src/main/java/feast/core/model/FeatureGroupInfo.java]: Do we really need to have oldStoreOpts here? Can this also be a method on StorageInfo?

**Comment by zhilingc** [comment]
Cleaned it up a bit.

File: core/src/main/java/feast/core/service/SpecService.java [hunk 4]
```diff
@@ -314,19 +344,30 @@ public EntityInfo registerEntity(EntitySpec spec) throws RegistrationException {
...
+        if (!storageInfo.getType().equals(spec.getType())
+            && !storageInfo
+                .getOptions()
+                .equals(TypeConversion.convertMapToJsonString(spec.getOptionsMap()))) {
+          throw new IllegalArgumentException("updating storage specs is not allowed");
+        }
+        return storageInfo;
+      } else {
+        storageInfo = new StorageInfo(spec);
+        StorageInfo out = storageInfoRepository.saveAndFlush(storageInfo);
```
Comment (woop) [inline core/src/main/java/feast/core/service/SpecService.java]: Isn't there a more descriptive variable name that we can use instead of `out`?

**Pull request merged**